itou/archive/management/commands/anonymize_professionals.py

```diff
@@ -41,6 +53,114 @@ def reset_notified_professionals_with_recent_activity(self):
 
+    @transaction.atomic
+    def anonymize_professionals_after_grace_period(self):
+        now = timezone.now()
+        grace_period_since = now - GRACE_PERIOD
+        self.logger.info("Archiving professionals after grace period, notified before: %s", grace_period_since)
+
+        users = self.get_users(grace_period_since)
+
+        # split users to anonymize into those that can be deleted and those that can only be anonymized
+        users_to_delete = self.get_users_to_anonymize_and_delete(users)
+
+        users_not_to_delete = [user for user in users if user not in users_to_delete]
+        users_to_anonymize = self.get_users_to_anonymize_without_deletion(users_not_to_delete)
+
+        # users to anonymize or delete that have an email set
+        users_to_remove_from_contact = [user for user in users if user.email]
+
+        if self.wet_run:
+            for user in users:
+                ArchiveUser(user).send()
+
+            self.anonymize_and_delete_professionals(users_to_delete)
+            self.anonymize_professionals_without_deletion(users_to_anonymize)
+            self.remove_from_contact(users_to_remove_from_contact)
+
+        self.logger.info("Anonymized professionals after grace period, count: %d", len(users))
+        self.logger.info(
+            "Included in this count: %d to delete, %d to remove from contact",
+            len(users_to_delete),
+            len(users_to_remove_from_contact),
+        )
+
+    def get_users(self, grace_period_since):
+        return list(
+            User.objects.filter(
+                kind__in=UserKind.professionals(), upcoming_deletion_notified_at__lte=grace_period_since
+            )
+            .annotate(is_deactivated=Q(email__isnull=True))
+            .order_by("is_deactivated", "upcoming_deletion_notified_at")
+            .select_for_update(of=("self",), skip_locked=True)[: self.batch_size]
+        )
+
+    def get_users_to_anonymize_and_delete(self, users):
+        related_objects_to_check = get_filter_kwargs_on_user_for_related_objects_to_check()
+        has_membership_in_authorized_organization_sqs = PrescriberMembership.objects.filter(
+            user_id=OuterRef("id"), organization__authorization_status=PrescriberAuthorizationStatus.VALIDATED
+        )
+        return list(
+            User.objects.filter(id__in=[user.id for user in users])
+            .filter(**related_objects_to_check)
+            .annotate(has_membership_in_authorized_organization=Exists(has_membership_in_authorized_organization_sqs))
+            .prefetch_related(
+                "companymembership_set",
+                "institutionmembership_set",
+                "prescribermembership_set",
+            )
+        )
+
+    def get_users_to_anonymize_without_deletion(self, users):
+        # we do not want to anonymize users that have no email set, as they are already anonymized
+        return list(User.objects.filter(id__in=[user.id for user in users], email__isnull=False))
```

Comment:
On pourrait éviter cette requête SQL (et l'appel à cette fonction) avec:
`users_to_anonymize = [user for user in users_not_to_delete and user.email is not None]`